connmgr/connmanager.go

```diff
@@ -0,0 +1,235 @@
+// Copyright (c) 2016 The btcsuite developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package connmgr
+
+import (
+	"net"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+// DialFunc defines a function that dials a connection.
+type DialFunc func(string, string) (net.Conn, error)
+
+// AddressFunc defines a function that returns a network address to connect to.
+type AddressFunc func() string
+
+// ConnState represents the state of the requested connection.
+type ConnState uint8
+
+// ConnState can be either pending, established, disconnected or failed.  When
+// a new connection is requested, it is attempted and categorized as
+// established or failed depending on the connection result.  An established
+// connection which was disconnected is categorized as disconnected.
+const (
+	ConnPending      ConnState = 0x0
+	ConnEstablished  ConnState = 0x1
+	ConnDisconnected ConnState = 0x2
+	ConnFailed       ConnState = 0x3
+)
+
+// ConnReq is the connection request to a network address. If permanent, the
+// connection will be retried on disconnection.
+type ConnReq struct {
+	Addr      string
+	Permanent bool
+
+	state      ConnState
+	retryCount uint32
+}
+
+// State is the connection state of the requested connection.
+func (c *ConnReq) State() ConnState {
+	return c.state
```

Comment:
This will race.  It can access the state variable from any goroutine and it's set in the `connHandler` goroutine below.